cpp/include/rapidsmpf/streaming/core/lineariser.hpp

```diff
@@ -0,0 +1,157 @@
+ * std::vector<Node> tasks;
+ * // Draining the lineariser will pull from all the input channels until they are
+ * // shutdown and send to the output channel until it is consumed.
+ * tasks.push_back(linearise->drain(ctx));
+ * for (auto& ch_in: lineariser->get_inputs()) {
+ *   // Each producer promises to send an increasing stream of sequence ids.
+ *   // For best performance they should attempt to take "interleaved" ids from a shared
+ *   // task list. That is, don't have producer-0 produce the first K ids, producer-1 the
+ *   // next K, and so forth.
+ *   tasks.push_back(producer(ctx, ch_in, ...));
+ * }
+ * coro_results(co_await coro::when_all(std::move(tasks)));
+ * // ch_out will see inputs in global total order of sequence id.
+ * @endcode
+ */
+class Lineariser {
+  public:
+    /**
+     * @brief Create a new `Lineariser` into an output channel.
+     *
+     * @param ch_out The output channel
+     * @param num_producers The number of producers.
+     */
+    Lineariser(std::shared_ptr<Channel> ch_out, std::size_t num_producers)
+        : ch_out_{std::move(ch_out)} {
+        inputs_.reserve(num_producers);
+        for (std::size_t i = 0; i < num_producers; i++) {
+            inputs_.push_back(std::make_shared<Channel>());
+        }
+    }
+
+    /**
+     * @brief Get a reference to the input channels.
+     *
+     * @return Reference to the `Channel`s to send into.
+     *
+     * @note Behaviour is undefined if more than one producer coroutine sends into the
+     * same channel.
+     */
+    std::vector<std::shared_ptr<Channel>>& get_inputs() {
+        return inputs_;
+    }
+
+    /**
+     * @brief Process inputs and send to the output channel.
+     *
+     * @param ctx The execution context to use.
+     * @return Coroutine representing the linearised sends of all producers.
+     *
+     * @note This coroutine should be awaited in a `coro::when_all` with all of the
+     * producer tasks.
+     */
+    Node drain(std::shared_ptr<Context> ctx) {
+        ShutdownAtExit c{ch_out_};
+        co_await ctx->executor()->schedule();
+        // Invariant: the heap always contains exactly zero-or-one messages from each
+        // producer. We always extract the minimum element, and then repoll the producer
+        // that made that element. First poll all producers.
+        for (std::size_t p = 0; p < inputs_.size(); p++) {
+            auto msg = co_await inputs_[p]->receive();
+            if (!msg.empty()) {
+                min_heap_.emplace(p, std::move(msg));
+            }
+        }
+        while (!min_heap_.empty()) {
+            auto item = min_heap_.pop_top();
+            co_await ch_out_->send(std::move(item.value));
+            // And refill from the producer we just consumed from.
+            auto msg = co_await inputs_[item.producer]->receive();
+            if (!msg.empty()) {
+                min_heap_.emplace(item.producer, std::move(msg));
+            }
+        }
```

Comment:
Here, the loop is held back by the slowest producer. It pushes back all the other input channels. Could this potentially cause a deadlock, like we saw in the Union case (which had multiple consumers though)?